### PR TITLE
Feature/debianize

### LINKED ISSUE #1199: [Feature] Add arm64 deb builds to main repo
## Detailed Description
Add arm64 deb builds to the main APT repo so that we can install the latest version on an Raspberry Pi or generic ARM system.

## Context
We are stuck with the defaults in the upstream repo otherwise.

### PR DESCRIPTION
**Pull request by neolynx**
Fixes #1199, #960, #1108, #1287, #934, #425

## Description of the Change

Debianize aptly and be compatible with official debian package. 

This PR improved building and packaging and also improvides the apt repository structure
- replace goxc and use go directly for debian builds
- debianize and use dpkg-buildpackage for building debian packages
- build for architectures: arm64, arm64, armhf, i386
- read go version from go.mod everywhere
- read package version from debian/changelog
- build FreeBSD, MacOS and Linux separately
- introduce new apt repository structure on repo.aptly.info
- support Ubuntu 24.04 (noble) and Debian testing (trixie) builds
- use go 1.22
- downgrade/upgrade go modules to match versions available in Debian/trixie (see [Google Spreadsheet](https://docs.google.com/spreadsheets/d/1HFo1sulC0csRTbjZAoBNVwnOq290Q7CCuOonXn7KWtU/edit?gid=1986176366#gid=1986176366))

## APT Repositories

Aptly is now available on repo.aptly.info for:
- Debian buster, bullseye, bookworm, trixie
- Ubuntu focal, jammy
- Architectures: arm64, arm64, armhf, i386

### Repo Key
```
sudo wget http://repo.aptly.info/pubkey.txt -O /etc/apt/keyrings/aptly.asc
```

### New CI repo:
```
deb [signed-by=/etc/apt/keyrings/aptly.asc] http://repo.aptly.info/ci bookworm main
```

### New Release repo:
```
deb [signed-by=/etc/apt/keyrings/aptly.asc] http://repo.aptly.info/release bookworm main
```

## TODO

- [x] cleanup debian package / service
   - [x] zsh completion
- [x] be compatible with official debian package (switching without loosing config/data)
- [x] CI prod release use dpkg-buildpackage
- [x] https://github.com/aptly-dev/aptly/issues/1108
- [x] check /etc/aptly.conf readability
- [x] check upgrade from debian version 1.5.0+ds1-1+b4
- [x] check upgrade from previous nightly builds
- [x] check contents in non deb builds are the same

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -89,81 +94,117 @@ jobs:
...
+          - name: "Debian 11"
             suite: bullseye
-            GOVER: '1.21'
-            install: "make ca-certificates git curl"
-          - name: Debian-12
-            image: "debian:bookworm"
-            suite: bookworm
-            install: "make ca-certificates git curl golang"
+            image: debian:bullseye
+          - name: "Ubuntu 22.04"
```
[the remark this answers is not in the document]
Comment (neolynx) [reply .github/workflows/ci.yml:122]: Good point, I added Ubuntu 24.04 noble build

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -30,39 +28,41 @@ jobs:
       BOTO_CONFIG: /dev/null
       GO111MODULE: "on"
       GOPROXY: "https://proxy.golang.org"
-      GOVER: '1.21'
+      GOVER: '1.19'
```
[the remark this answers is not in the document]
Comment (neolynx) [reply .github/workflows/ci.yml]: i removed the hardcoding everywhere now. 

Thanks for the review !

**Comment by iofq** [comment]
I have a few commits in https://github.com/iofq/aptly master branch that may fit well. 

Namely added release workflow to gitlab CI (which means we can completely remove goxc as I think it wasn't quite working for this anyway)

**Comment by neolynx** [comment]
oh, great. goxc is still in bcs of the MacOS, FreeBSD and legacy linux builds... 
have you been able to build them without goxc ?

**Comment by iofq** [comment]
Yep here is an example release: https://github.com/iofq/aptly/releases/tag/v1.15.0

Still cleaning it up and trying to figure out a changelog

File: .github/workflows/ci.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (neolynx) [reply .github/workflows/ci.yml:125]: also added debian/testing (trixie)

**Comment by iofq** [comment]
Opened https://github.com/aptly-dev/aptly/pull/1340 to pull my changes into this branch.

Here is an example release: https://github.com/iofq/aptly/releases/tag/v1.22.0

There should also be an auto-generated changelog on the release containing all PRs since the last release, however that's difficult to test

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -89,81 +93,168 @@ jobs:
...
+
+      - name: Check is CI or Release
+        run: |
+          if [ -z "`git tag --points-at HEAD`" ]; then
+            releasetype=ci
+          else
+            releasetype=release
+          fi
+          echo RELEASETYPE=$releasetype >> $GITHUB_OUTPUT
+        id: releasetype
```
Comment (iofq) [inline .github/workflows/ci.yml:163]: this should probably also be moved to the Makefile, can also be merged with the `RELEASE` variable I added there for a similar purpose

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -89,81 +93,168 @@ jobs:
...
+          if [ "${{ steps.releasetype.outputs.RELEASETYPE }}" = "ci" ]; then
+            DEBEMAIL="CI <runner@github>" dch -v `make version` "CI build"
+          fi
+          buildtype="any"
+          if [ "${{ matrix.arch }}" = "amd64" ]; then  # build 'all' type package only on amd64
+            buildtype="any,all"
+          fi
+          dpkg-buildpackage -us -uc --build=$buildtype -d --host-arch=${{ matrix.arch }}
+          mkdir -p build && mv ../*.deb build/
+          cd build && ls -l *.deb
```
Comment (iofq) [inline .github/workflows/ci.yml:179]: can we merge this with system/build-deb at all, or add it to the Makefile?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -89,81 +93,168 @@ jobs:
...
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ steps.goversion.outputs.GOVER }}
+
+      - name: "Build aptly ${{ matrix.goos }}/${{ matrix.goarch }}"
+        env:
+          GOBIN: /usr/local/bin
+        run: |
+          make binaries GOOS=${{ matrix.goos }} GOARCH=${{ matrix.goarch }}
+          cd build; unzip aptly_${{ steps.releaseversion.outputs.VERSION }}_${{ matrix.goos }}_${{ matrix.goarch }}.zip
```
Comment (iofq) [inline .github/workflows/ci.yml:237]: I mentioned elsewhere, but this is breaking the archives actually uploading in the gh-release step. The double-zip only appears in the github actions output; once you upload the zips to a release, they are formatted correctly.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (neolynx) [reply .github/workflows/ci.yml]: moved it to Makefile

**Comment by neolynx** [comment]
I think CI and release build are OK now :)

However, since this is a PR and not on master branch, could you maybe test in your [iofq:master](https://github.com/iofq/aptly/tree/master) branch if the correct version is built, debian release upload and gh-release are running ?

Thanks in advance !

**Comment by iofq** [comment]
https://github.com/iofq/aptly/releases/tag/v1.41.0

LGTM!

**Comment by neolynx** [comment]
thanks a lot !

I added some more checks to not publish a release version to ci and prod...

if you approve, we can merge :+1:

**Review by iofq — APPROVED** [review]

**Pull request merged**